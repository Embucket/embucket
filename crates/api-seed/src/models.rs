use fake::locales::Data;
use fake::faker::{name::raw::Name, lorem::en::Word};
use fake::{Dummy, Fake, locales::EN};
use serde::{Deserialize, Serialize};
use std::marker::PhantomData;
use core_metastore::VolumeType;

///// Traits

pub trait Generator<T> {
    fn generate(&self, index: usize) -> T;
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct WithCount<T, G>
where
    G: Generator<T>,
{
    count: usize,
    generator: G,
    #[serde(skip)]
    _marker: PhantomData<T>,
}

impl<T, G> WithCount<T, G>
where
    G: Generator<T>,
{
    pub fn new(count: usize, generator: G) -> Self {
        Self { count, generator, _marker: PhantomData }
    }

    pub fn generate(&self) -> Vec<T> {
        (0..self.count)
            .map(|i| self.generator.generate(i))
            .collect()
    }
}

///// Super Volume

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum SuperVolumeType {
    Volume(Volume),
    VolumeGenerator(VolumeGenerator),
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SuperVolume {
    #[serde(flatten)]
    pub volume: SuperVolumeType,
}

impl SuperVolume {
    pub fn materialize(self) -> Volume {
        match self.volume {
            SuperVolumeType::VolumeGenerator(volume_generator) => volume_generator.generate(0),
            SuperVolumeType::Volume(volume) => volume,
        }
    }
}

///// Volume

#[derive(Debug, Serialize, Deserialize)]
pub struct Volume {
    pub volume_name: String,
    pub databases: Vec<Database>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct VolumeGenerator {
    databases_gen: WithCount<Database, DatabaseGenerator>,
    volume_name: Option<String>, // if None value will be generated    
}

impl Generator<Volume> for VolumeGenerator {
    fn generate(&self, _index: usize) -> Volume {
        Volume {
            databases: self.databases_gen.generate(),
            volume_name: self.volume_name.clone().unwrap_or_else(|| Name(EN).fake()),
        }
    }
}

///// Database

#[derive(Debug, Serialize, Deserialize)]
pub struct Database {
    name: String,
    schemas: Vec<Schema>,
}

#[derive(Debug, Serialize, Deserialize)]
struct DatabaseGenerator {
    name: Option<String>, // if None value will be generated
    schemas_gen: WithCount<Schema, SchemaGenerator>,
}

impl Generator<Database> for DatabaseGenerator {
    fn generate(&self, _index: usize) -> Database {
        Database {
            name: Name(EN).fake(),
            schemas: self.schemas_gen.generate(),
        }
    }
}


///// Schema

#[derive(Debug, Serialize, Deserialize)]
pub struct Schema {
    name: String,
    tables: Vec<Table>,
}

#[derive(Debug, Serialize, Deserialize)]
struct SchemaGenerator {
    name: Option<String>, // if None value will be generated
    tables_gen: WithCount<Table, TableGenerator>,
}

impl Generator<Schema> for SchemaGenerator {
    fn generate(&self, _index: usize) -> Schema {
        Schema {
            name: Name(EN).fake(),
            tables: self.tables_gen.generate(),
        }
    }
}


///// Table

#[derive(Debug, Serialize, Deserialize)]
pub struct Table {
    name: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct TableGenerator {
    name: Option<String>, // if None value will be generated
    columns_gen: WithCount<Column, ColumnGenerator>,
}

impl Generator<Table> for TableGenerator {
    fn generate(&self, _index: usize) -> Table {
        Table {
            name: Name(EN).fake(),
        }
    }
}


///// Column

#[derive(Debug, Serialize, Deserialize)]
struct Column {
    name: String,
    #[serde(rename = "type")]
    col_type: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct ColumnGenerator {
    name: Option<String>, // if None value will be generated
}

impl Generator<Column> for ColumnGenerator {
    fn generate(&self, _index: usize) -> Column {
        Column {
            name: Name(EN).fake(),
            col_type: "".to_string(),
        }
    }
}


    // let table_gen = TableGenerator;
    // let schema_gen = SchemaGenerator {
    //     table_gen: WithCount::new(3, table_gen),
    // };
    // let db_gen = DatabaseGenerator {
    //     schema_gen: WithCount::new(2, schema_gen),
    // };

    // let dbs = WithCount::new(1, db_gen).generate();


// #[derive(Debug, Serialize, Deserialize)]
// struct SeedFactor {
//     pub name: String,
//     pub x_min: u16,
//     pub x_max: u16,
//     pub rng_seed: Option<u64>,
// }
