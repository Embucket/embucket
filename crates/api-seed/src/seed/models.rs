use core_metastore::VolumeType;
use fake::faker::{lorem::en::Word, name::raw::Name};
use fake::{Fake, locales::EN};
use serde::{Deserialize, Serialize};
use std::marker::PhantomData;

struct FakeProvider;

impl FakeProvider {
    pub fn volume_type() -> VolumeType {
        VolumeType::Memory // no random volume type
    }
    pub fn person_name() -> String {
        Name(EN).fake()
    }
    pub fn entity_name() -> String {
        let one: String = Word().fake();
        let two: String = Word().fake();
        format!("{}_{}", one.to_lowercase(), two.to_lowercase())
    }
}

///// Traits

pub trait Generator<T> {
    fn generate(&self, index: usize) -> T;
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct WithCount<T, G>
where
    G: Generator<T>,
{
    count: usize,
    generator: G,
    #[serde(skip)]
    _marker: PhantomData<T>,
}

impl<T, G> WithCount<T, G>
where
    G: Generator<T>,
{
    pub fn new(count: usize, generator: G) -> Self {
        Self {
            count,
            generator,
            _marker: PhantomData,
        }
    }

    pub fn generate(&self) -> Vec<T> {
        (0..self.count)
            .map(|i| self.generator.generate(i))
            .collect()
    }
}

///// Seed Root

#[derive(Debug, Serialize, Deserialize)]
pub struct SeedRoot {
    pub volumes: Vec<SuperVolume>,
}

///// Super Volume

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum SuperVolumeType {
    Volume(Volume),
    VolumeGenerator(VolumeGenerator),
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SuperVolume {
    #[serde(flatten)]
    pub volume: SuperVolumeType,
}

impl SuperVolume {
    pub fn materialize(self) -> Volume {
        match self.volume {
            SuperVolumeType::VolumeGenerator(volume_generator) => volume_generator.generate(0),
            SuperVolumeType::Volume(volume) => volume,
        }
    }
}

///// Volume

#[derive(Debug, Serialize, Deserialize)]
pub struct Volume {
    pub volume_name: String,
    pub volume_type: VolumeType,
    pub databases: Vec<Database>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct VolumeGenerator {
    pub volume_name: Option<String>, // if None value will be generated
    pub volume_type: Option<VolumeType>,
    pub databases_gen: WithCount<Database, DatabaseGenerator>,
}

impl Generator<Volume> for VolumeGenerator {
    fn generate(&self, _index: usize) -> Volume {
        Volume {
            volume_name: self
                .volume_name
                .clone()
                .unwrap_or_else(|| FakeProvider::entity_name()),
            volume_type: self.volume_type.clone().unwrap_or(VolumeType::Memory),
            databases: self.databases_gen.generate(),
        }
    }
}

///// Database

#[derive(Debug, Serialize, Deserialize)]
pub struct Database {
    name: String,
    schemas: Vec<Schema>,
}

#[derive(Debug, Serialize, Deserialize)]
struct DatabaseGenerator {
    name: Option<String>, // if None value will be generated
    schemas_gen: WithCount<Schema, SchemaGenerator>,
}

impl Generator<Database> for DatabaseGenerator {
    fn generate(&self, _index: usize) -> Database {
        Database {
            name: FakeProvider::entity_name(),
            schemas: self.schemas_gen.generate(),
        }
    }
}

///// Schema

#[derive(Debug, Serialize, Deserialize)]
pub struct Schema {
    name: String,
    tables: Vec<Table>,
}

#[derive(Debug, Serialize, Deserialize)]
struct SchemaGenerator {
    name: Option<String>, // if None value will be generated
    tables_gen: WithCount<Table, TableGenerator>,
}

impl Generator<Schema> for SchemaGenerator {
    fn generate(&self, _index: usize) -> Schema {
        Schema {
            name: FakeProvider::entity_name(),
            tables: self.tables_gen.generate(),
        }
    }
}

///// Table

#[derive(Debug, Serialize, Deserialize)]
pub struct Table {
    name: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct TableGenerator {
    name: Option<String>, // if None value will be generated
    columns_gen: WithCount<Column, ColumnGenerator>,
}

impl Generator<Table> for TableGenerator {
    fn generate(&self, _index: usize) -> Table {
        Table {
            name: FakeProvider::entity_name(),
        }
    }
}

///// Column

#[derive(Debug, Serialize, Deserialize)]
struct Column {
    name: String,
    #[serde(rename = "type")]
    col_type: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct ColumnGenerator {
    name: Option<String>, // if None value will be generated
}

impl Generator<Column> for ColumnGenerator {
    fn generate(&self, _index: usize) -> Column {
        Column {
            name: FakeProvider::entity_name(),
            col_type: "".to_string(),
        }
    }
}

// let table_gen = TableGenerator;
// let schema_gen = SchemaGenerator {
//     table_gen: WithCount::new(3, table_gen),
// };
// let db_gen = DatabaseGenerator {
//     schema_gen: WithCount::new(2, schema_gen),
// };

// let dbs = WithCount::new(1, db_gen).generate();

// #[derive(Debug, Serialize, Deserialize)]
// struct SeedFactor {
//     pub name: String,
//     pub x_min: u16,
//     pub x_max: u16,
//     pub rng_seed: Option<u64>,
// }
