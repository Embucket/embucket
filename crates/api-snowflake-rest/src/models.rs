cfg_if::cfg_if! {
    if #[cfg(feature = "default-server")] {
        use core_executor::models::ColumnInfo as ColumnInfoModel;
        use core_history::result_set::Row;
    } else {
        // Define simple representation for Row
        // As with external-server we won't have any dependences (especially trivial)
        type Row = Vec<serde_json::Value>;
    }
}

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use uuid::Uuid;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LoginRequestQueryParams {
    #[serde(rename = "request_id")]
    pub request_id: String,
    pub database_name: Option<String>,
    pub schema_name: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct LoginRequestBody {
    pub data: LoginRequestData,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct LoginResponse {
    pub data: Option<LoginResponseData>,
    pub success: bool,
    pub message: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct LoginResponseData {
    pub token: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "UPPERCASE")]
pub struct LoginRequestData {
    pub client_app_id: String,
    pub client_app_version: String,
    pub svn_revision: Option<String>,
    pub account_name: String,
    pub login_name: String,
    pub client_environment: ClientEnvironment,
    pub password: String,
    pub session_parameters: HashMap<String, serde_json::Value>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "UPPERCASE")]
pub struct ClientEnvironment {
    pub application: String,
    pub os: String,
    pub os_version: String,
    pub python_version: String,
    pub python_runtime: String,
    pub python_compiler: String,
    pub ocsp_mode: String,
    pub tracing: u32,
    pub login_timeout: Option<u32>,
    pub network_timeout: Option<u32>,
    pub socket_timeout: Option<u32>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryRequest {
    pub request_id: Uuid,
    pub retry_count: u16,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryRequestBody {
    pub sql_text: String,
    pub async_exec: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AbortRequestBody {
    pub sql_text: String,
    pub request_id: Uuid, // duplicate in body, taken from snowflake connector
}

#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct ResponseData {
    #[serde(rename = "rowtype")]
    pub row_type: Vec<ColumnInfo>,
    #[serde(rename = "rowsetBase64")]
    pub row_set_base_64: Option<String>,
    #[serde(rename = "rowset")]
    pub row_set: Option<Vec<Row>>,
    pub total: Option<u32>,
    #[serde(rename = "queryResultFormat")]
    pub query_result_format: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sql_state: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub query_id: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct JsonResponse {
    pub data: Option<ResponseData>,
    pub success: bool,
    pub message: Option<String>,
    pub code: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ColumnInfo {
    name: String,
    database: String,
    schema: String,
    table: String,
    nullable: bool,
    #[serde(rename = "type")]
    r#type: String,
    #[serde(rename = "byteLength")]
    byte_length: Option<i32>,
    length: Option<i32>,
    scale: Option<i32>,
    precision: Option<i32>,
    collation: Option<String>,
}

#[cfg(feature = "default-server")]
impl From<ColumnInfoModel> for ColumnInfo {
    fn from(column_info: ColumnInfoModel) -> Self {
        Self {
            name: column_info.name,
            database: column_info.database,
            schema: column_info.schema,
            table: column_info.table,
            nullable: column_info.nullable,
            r#type: column_info.r#type,
            byte_length: column_info.byte_length,
            length: column_info.length,
            scale: column_info.scale,
            precision: column_info.precision,
            collation: column_info.collation,
        }
    }
}

#[derive(Clone, Default)]
pub struct Auth {
    pub demo_user: String,
    pub demo_password: String,
}
