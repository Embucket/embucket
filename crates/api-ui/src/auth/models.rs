use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

#[derive(Serialize, Clone, PartialEq, Eq, ToSchema)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(test, derive(Debug, Deserialize))]
pub struct RefreshTokenResponse {
    pub access_token: String,
}

#[derive(Deserialize, Clone, PartialEq, Eq, ToSchema)]
#[cfg_attr(test, derive(Serialize))]
pub struct LoginPayload {
    pub username: String,
    pub password: String,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct AccountResponse {
    pub username: String,
}

// For internal use, should not be included into open api spec
#[derive(Serialize, Deserialize)]
#[cfg_attr(test, derive(Debug))]
pub struct Claims {
    pub sub: String, // token issued to a particular user
    pub aud: String, // validate audience since as it can be deployed on multiple hosts
    pub iat: i64,    // Issued At
    pub exp: i64,    // Expiration Time
}
