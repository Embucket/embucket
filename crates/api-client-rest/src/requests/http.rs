#![allow(clippy::unwrap_used, clippy::expect_used)]

use super::error::HttpRequestError;
use http::{HeaderMap, HeaderValue, Method, StatusCode, header};
use reqwest;
use std::fmt::Display;

#[derive(Debug)]
pub struct HttpErrorData {
    pub method: Method,
    pub url: String,
    pub headers: HeaderMap<HeaderValue>,
    pub status: StatusCode,
    pub body: String,
    pub error: HttpRequestError,
}

impl From<HttpErrorData> for HttpRequestError {
    fn from(value: HttpErrorData) -> Self {
        let HttpErrorData { error, status, .. } = value;
        HttpRequestError::HttpRequest {
            message: error.to_string(),
            status,
        }
    }
}

impl Display for HttpErrorData {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.error)
    }
}

/// As of minimalistic interface this doesn't support checking request/response headers
pub async fn http_req_with_headers<T: serde::de::DeserializeOwned>(
    client: &reqwest::Client,
    method: Method,
    headers: HeaderMap,
    url: &str,
    payload: String,
) -> Result<(HeaderMap, T), HttpErrorData> {
    let res = client
        .request(method.clone(), url)
        .headers(headers)
        .body(payload)
        .send()
        .await;

    let response = res.unwrap();
    if response.status() == StatusCode::OK {
        let headers = response.headers().clone();
        let status = response.status();
        let text = response.text().await.expect("Failed to get response text");
        if text.is_empty() {
            // If no actual type retuned we emulate unit, by "null" value in json
            Ok((
                headers,
                serde_json::from_str::<T>("null").expect("Failed to parse response"),
            ))
        } else {
            let json = serde_json::from_str::<T>(&text);
            match json {
                Ok(json) => Ok((headers, json)),
                Err(err) => {
                    // Normally we don't expect error here, and only have http related error to return
                    Err(HttpErrorData {
                        method,
                        url: url.to_string(),
                        headers,
                        status,
                        body: text,
                        error: HttpRequestError::HttpRequest {
                            message: err.to_string(),
                            status,
                        },
                    })
                }
            }
        }
    } else {
        let error = response
            .error_for_status_ref()
            .expect_err("Expected error, http code not OK");
        // Return custom error as reqwest error has no body contents
        Err(HttpErrorData {
            method,
            url: url.to_string(),
            headers: response.headers().clone(),
            status: response.status(),
            error: HttpRequestError::HttpRequest {
                message: error.to_string(),
                status: response.status(),
            },
            body: response.text().await.expect("Failed to get response text"),
        })
    }
}

/// As of minimalistic interface this doesn't support checking request/response headers
pub async fn http_req<T: serde::de::DeserializeOwned>(
    client: &reqwest::Client,
    method: Method,
    url: &str,
    payload: String,
) -> Result<T, HttpErrorData> {
    let headers = HeaderMap::from_iter(vec![(
        header::CONTENT_TYPE,
        HeaderValue::from_static("application/json"),
    )]);
    let (_, res) = http_req_with_headers(client, method, headers, url, payload).await?;
    Ok(res)
}
