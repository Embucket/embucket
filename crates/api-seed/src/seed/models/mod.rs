pub mod volume;
pub mod schema;
pub mod database;
pub mod table;
pub mod column;

pub use volume::*;
pub use schema::*;
pub use database::*;
pub use table::*;
pub use column::*;

use serde::{Deserialize, Serialize};
use std::marker::PhantomData;

#[derive(Debug, Serialize, Deserialize)]
pub struct VolumesRoot {
    pub volumes: Vec<VolumeGenerator>,
}

impl VolumesRoot {
    pub fn generate(&self) -> Vec<Volume> {
        self.volumes
            .iter()
            .enumerate()
            .map(|(i, v)| v.generate(i))
            .collect()
    }
}

pub trait Generator<T> {
    // create entity, item index is just for reference
    fn generate(&self, index: usize) -> T;
}

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct WithCount<T, G>
where
    G: Generator<T>,
{
    count: usize,
    template: G,
    #[serde(skip)]
    _marker: PhantomData<T>,
}

impl<T, G> WithCount<T, G>
where
    G: Generator<T>,
{
    pub fn new(count: usize, template: G) -> Self {
        Self {
            count,
            template,
            _marker: PhantomData,
        }
    }

    // create items for template, item index is just for reference
    pub fn vec_with_count(&self, index: usize) -> Vec<T> {
        // call generate n times
        (0..self.count)
            .map(|i| self.template.generate(i))
            .collect()
    }
}
