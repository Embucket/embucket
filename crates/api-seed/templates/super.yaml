volume_name:
databases_gen:
  count: 1
  generator:
    name: null
    schemas_gen:
      count: 1
      generator:
        name: null
        tables_gen:
          count: 1
          generator:
            name: null
            columns_gen:
              count: 1
              generator:
                name: null
