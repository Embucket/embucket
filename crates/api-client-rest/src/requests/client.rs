#![allow(clippy::expect_used)]
use super::error::{HttpRequestError, HttpRequestResult, InvalidHeaderValueSnafu, SerializeSnafu};
use super::helpers::get_set_cookie_name_value_map;
use super::http::http_req_with_headers;
use api_structs::auth::{AuthResponse, LoginPayload};
use api_structs::query::QueryCreatePayload;
use http::{HeaderMap, HeaderValue, Method, StatusCode, header};
use reqwest;
use serde::de::DeserializeOwned;
use serde_json::json;
use snafu::ResultExt;
use std::net::SocketAddr;

pub struct AuthenticatedClient {
    client: reqwest::Client,
    addr: SocketAddr,
    access_token: String,
    refresh_token: String,
    session_id: Option<String>,
}

#[async_trait::async_trait]
pub trait AuthenticatedRequests {
    fn addr(&self) -> SocketAddr;

    /// should login before run queries
    async fn login(&mut self, user: &str, password: &str) -> HttpRequestResult<AuthResponse>;

    async fn refresh(&mut self) -> HttpRequestResult<AuthResponse>;

    async fn query<T: DeserializeOwned + Send>(&mut self, query: &str) -> HttpRequestResult<T>
    where
        Self: Sized;

    async fn generic_request<I, T>(
        &mut self,
        method: Method,
        url: &str,
        payload: &I,
    ) -> HttpRequestResult<T>
    where
        I: serde::Serialize + Sync,
        T: serde::de::DeserializeOwned + Send;
}

impl AuthenticatedClient {
    #[must_use]
    pub fn new(addr: SocketAddr) -> Self {
        Self {
            client: reqwest::Client::new(),
            addr,
            access_token: String::new(),
            refresh_token: String::new(),
            session_id: None,
        }
    }

    fn set_tokens_from_auth_response(&mut self, headers: &HeaderMap, auth_response: &AuthResponse) {
        let from_set_cookies = get_set_cookie_name_value_map(headers);
        if let Some(refresh_token) = from_set_cookies.get("refresh_token") {
            self.refresh_token.clone_from(refresh_token);
        }
        self.access_token.clone_from(&auth_response.access_token);
    }

    fn set_session_id_from_response_headers(&mut self, headers: &HeaderMap) {
        let from_set_cookies = get_set_cookie_name_value_map(headers);
        self.session_id = from_set_cookies.get("id").cloned();
    }

    async fn generic_request_no_refresh<
        I: serde::Serialize + Sync,
        T: serde::de::DeserializeOwned + Send,
    >(
        &mut self,
        method: Method,
        url: &str,
        payload: &I,
    ) -> HttpRequestResult<T>
    where
        Self: Sized,
    {
        let Self {
            access_token,
            refresh_token,
            client,
            ..
        } = self;

        let mut headers = HeaderMap::from_iter(vec![
            (
                header::CONTENT_TYPE,
                HeaderValue::from_static("application/json"),
            ),
            (
                header::AUTHORIZATION,
                HeaderValue::from_str(format!("Bearer {access_token}").as_str())
                    .expect("Can't convert to HeaderValue"),
            ),
        ]);

        // prepare cookies
        let mut cookies = Vec::new();
        if !refresh_token.is_empty() {
            cookies.push(format!("refresh_token={refresh_token}"));
        }
        if let Some(session_id) = &self.session_id {
            cookies.push(format!("id={session_id}"));
        }
        if !cookies.is_empty() {
            headers.insert(
                header::COOKIE,
                HeaderValue::from_str(cookies.join(";").as_str())
                    .context(InvalidHeaderValueSnafu)?,
            );
        }

        let res = http_req_with_headers::<T>(
            client,
            method,
            headers,
            url,
            serde_json::to_string(&payload).context(SerializeSnafu)?,
        )
        .await
        .map_err(HttpRequestError::from);

        match res {
            Ok((headers, resp_data)) => {
                self.set_session_id_from_response_headers(&headers);
                Ok(resp_data)
            }
            Err(err) => Err(err),
        }
    }
}

#[async_trait::async_trait]
impl AuthenticatedRequests for AuthenticatedClient {
    fn addr(&self) -> SocketAddr {
        self.addr
    }

    async fn login(&mut self, user: &str, password: &str) -> HttpRequestResult<AuthResponse> {
        let Self { client, addr, .. } = self;

        let login_result = http_req_with_headers::<AuthResponse>(
            client,
            Method::POST,
            HeaderMap::from_iter(vec![(
                header::CONTENT_TYPE,
                HeaderValue::from_static("application/json"),
            )]),
            &format!("http://{addr}/ui/auth/login"),
            json!(LoginPayload {
                username: user.to_string(),
                password: password.to_string(),
            })
            .to_string(),
        )
        .await;

        match login_result {
            Ok((headers, auth_response)) => {
                self.set_tokens_from_auth_response(&headers, &auth_response);
                Ok(auth_response)
            }
            Err(err) => Err(HttpRequestError::from(err)),
        }
    }

    async fn refresh(&mut self) -> HttpRequestResult<AuthResponse> {
        let Self {
            client,
            addr,
            refresh_token,
            ..
        } = self;

        let refresh_result = http_req_with_headers::<AuthResponse>(
            client,
            Method::POST,
            HeaderMap::from_iter(vec![
                (
                    header::CONTENT_TYPE,
                    HeaderValue::from_static("application/json"),
                ),
                (
                    header::COOKIE,
                    HeaderValue::from_str(format!("refresh_token={refresh_token}").as_str())
                        .expect("Can't convert to HeaderValue"),
                ),
            ]),
            &format!("http://{addr}/ui/auth/refresh"),
            String::new(),
        )
        .await;

        match refresh_result {
            Ok((headers, auth_response)) => {
                self.set_tokens_from_auth_response(&headers, &auth_response);
                Ok(auth_response)
            }
            Err(err) => Err(HttpRequestError::from(err)),
        }
    }

    // sets access_token at refresh if expired
    async fn query<T: DeserializeOwned + Send>(&mut self, query: &str) -> HttpRequestResult<T>
    where
        Self: Sized,
    {
        let url = format!("http://{}/ui/queries", self.addr);
        let query_payload = QueryCreatePayload {
            worksheet_id: None,
            query: query.to_string(),
            context: None,
        };

        self.generic_request(Method::POST, &url, &query_payload)
            .await
    }

    async fn generic_request<I, T>(
        &mut self,
        method: Method,
        url: &str,
        payload: &I,
    ) -> HttpRequestResult<T>
    where
        I: serde::Serialize + Sync,
        T: serde::de::DeserializeOwned + Send,
    {
        match self
            .generic_request_no_refresh(method.clone(), url, payload)
            .await
        {
            Ok(t) => Ok(t),
            Err(HttpRequestError::HttpRequest {
                status: StatusCode::UNAUTHORIZED,
                ..
            }) => {
                let _refresh_resp = self.refresh().await?;
                self.generic_request_no_refresh(method, url, payload).await
            }
            Err(err) => Err(err),
        }
    }
}
