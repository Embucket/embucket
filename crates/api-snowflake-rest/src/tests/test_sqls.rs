#[cfg(not(feature = "external-server"))]
use crate::server::test_server::run_test_server;

// External server should be already running, we just return its address
#[cfg(feature = "external-server")]
use crate::tests::external_server::run_test_server;

use super::client::{get_query_result, login, query};
use crate::models::{JsonResponse, LoginResponse};
use http::header;
use std::net::SocketAddr;

fn insta_replace_filiters() -> Vec<(&'static str, &'static str)> {
    vec![(
        r"[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}",
        "UUID",
    )]
}

const DEMO_USER: &str = "embucket";
const DEMO_PASSWORD: &str = "embucket";

#[macro_export]
macro_rules! sql_test {
    ($name:ident, $sql:expr, $meta:expr) => {
        #[tokio::test]
        async fn $name() {
            let meta = $meta.lines()
                    .map(|line| line.trim_start())
                    .collect::<Vec<_>>()
                    .join("\n");
            let server_addr = run_test_server(DEMO_USER, DEMO_PASSWORD).await;
            let (snapshot, query_record) = snow_sql(&server_addr, $sql).await;

            let query_id = &snapshot.data.as_ref().map_or_else(
                || { format!("No data") },
                |data| { format!(
                    "Query UUID: {}",
                    data.query_id.clone().unwrap_or("No query ID".to_string())
                )},
            );

            let historical_codes = if let Some(data) = &query_record.data {
                format!("historical codes: sqlState: {}, errorCode: {}",
                    data.sql_state.as_ref().unwrap_or(&"".to_string()),
                    data.error_code.as_ref().unwrap_or(&"".to_string()))
            } else {
                "No data".to_string()
            };

            insta::with_settings!({
                snapshot_path => "snapshots",
                description => format!("{}\n{}\n{}", $sql, query_id, meta),
                sort_maps => true,
                filters => insta_replace_filiters(),
                info => &historical_codes,
            }, {
                insta::assert_json_snapshot!(snapshot);
            });
        }
    };
}

// Objectives: test error_code, sql_status, error message, result data.
// with insta snapshots
pub async fn snow_sql(server_addr: &SocketAddr, sql: &str) -> (JsonResponse, JsonResponse) {
    let client = reqwest::Client::new();
    let (headers, login_res) =
        login::<LoginResponse>(&client, server_addr, DEMO_USER, DEMO_PASSWORD)
            .await
            .expect("Failed to login");
    assert_eq!(headers.get(header::WWW_AUTHENTICATE), None);

    let access_token = login_res
        .data
        .clone()
        .map_or_else(String::new, |data| data.token);

    let mut async_exec = false;
    let mut sql = sql;
    // if sql ends with ;> it is async query
    if sql.ends_with(";>") {
        async_exec = true;
        sql = sql.trim_end_matches(";>");
    }
    let (_headers, res) =
        query::<JsonResponse>(&client, server_addr, &access_token, sql, async_exec)
            .await
            .expect("Failed to run query");

    let query_id = if let Some(data) = &res.data {
        data.query_id.clone().unwrap_or_default()
    } else {
        "No query ID".to_string()
    };
    let (_headers, history) =
        get_query_result::<JsonResponse>(&client, server_addr, &access_token, &query_id)
            .await
            .expect("Failed to get query result");

    (res, history)
}

mod snowflake_compatibility {
    use super::*;

    sql_test!(
        create_table_bad_syntax,
        "create table foo",
        "Snowflake:
        001003 (42000): UUID: SQL compilation error: 
        syntax error line 1 at position 16 unexpected '<EOF>'."
    );

    // incorrect sql_state: 42000, should be: 42502
    sql_test!(
        select_from_missing_table,
        "select * from foo",
        "Snowflake:
        002003 (42S02): UUID: SQL compilation error:
        Object 'FOO' does not exist or not authorized."
    );

    sql_test!(
        show_schemas_in_missing_db,
        "show schemas in database foo",
        "Snowflake:
        002043 (02000): UUID: SQL compilation error:
        Object does not exist, or operation cannot be performed."
    );

    sql_test!(
        select_1,
        "select 1",
        "Snowflake: 
        +---+
        | 1 |
        |---|
        | 1 |
        +---+"
    );

    sql_test!(
        select_1_async,
        "select 1;>",
        "Snowflake: scheduled query ID"
    );
}
