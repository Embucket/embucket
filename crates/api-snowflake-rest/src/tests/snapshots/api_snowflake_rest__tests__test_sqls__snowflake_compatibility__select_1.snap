---
source: crates/api-snowflake-rest/src/tests/test_sqls.rs
description: "select 1\nSnowflake: \n+---+\n| 1 |\n|---|\n| 1 |\n+---+"
expression: snapshot
info: "historical codes: sqlState: 02000, errorCode: "
---
{
  "data": {
    "rowtype": [
      {
        "name": "Int64(1)",
        "database": "",
        "schema": "",
        "table": "",
        "nullable": false,
        "type": "fixed",
        "byteLength": null,
        "length": null,
        "scale": 0,
        "precision": 38,
        "collation": null
      }
    ],
    "rowsetBase64": null,
    "rowset": [
      [
        1
      ]
    ],
    "total": 1,
    "queryResultFormat": "json",
    "sqlState": "02000",
    "queryId": "UUID"
  },
  "success": true,
  "message": "successfully executed",
  "code": "000200"
}
